syntax = "proto3";

service IDMService {
    rpc Login(LoginRequest) returns (LoginResponse);
    
    rpc ValidateToken(ValidateTokenRequest) returns (ValidateTokenResponse);
    
    rpc Logout(LogoutRequest) returns (LogoutResponse);

    rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);

    rpc ChangePassword(ChangePasswordRequest) returns (ChangePasswordResponse);

    rpc Register(RegisterRequest) returns (RegisterResponse);

    rpc DeleteAccount(DeleteAccountRequest) returns (DeleteAccountResponse);
}

message LoginRequest {
    string username = 1;
    string password = 2;
}

message LoginResponse {
    string token = 1;
}

message ValidateTokenRequest {
    string token = 1;
}

message ValidateTokenResponse {
    bool valid = 1;
    string username = 2;
    string role = 3;
}

message LogoutRequest {
    string token = 1;
}

message LogoutResponse {
    bool success = 1;
    string message = 2;
}

message ChangePasswordRequest {
    string username = 1;
    string current_password = 2;
    string new_password = 3;
    string token = 4;
}

message ChangePasswordResponse {
    bool success = 1;
    string message = 2;
}

message DeleteUserRequest{
    string username = 1;
}

message DeleteUserResponse{
    bool success = 1;
    string message = 2;
}

message RegisterRequest{
    string username = 1;
    string password = 2;
}

message RegisterResponse{
    bool success = 1;
    string message = 2;
}

message DeleteAccountRequest{
    string username = 1;
}

message DeleteAccountResponse{
    bool success = 1;
    string message = 2;
}